test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java

```diff
@@ -88,10 +96,120 @@ public void testEncoding() throws Exception {
+            arg = new HTTPArgument("name.?", encodedValue, true, "UTF-8");
+            assertNotEquals("arg.getValue() should not be equal to " + value, value, arg.getValue());
+        }
+    
+        private void testEncodings(String encoding) {
+            HTTPArgument arg;
+            arg = new HTTPArgument("name.?", "value_ here", false, encoding);
+            assertEquals("arg.getName() should be equal to name.?", "name.?", arg.getName());
+            assertEquals("arg.getValue() should be equal to value_here", "value_ here", arg.getValue());
+            assertEquals("arg.getEncodedName() should be equal to name.%3F", "name.%3F", arg.getEncodedName());
+            assertEquals("arg.getValue() should be equal to value_+here", "value_+here", arg.getEncodedValue());
+            // Show that can bypass encoding:
+            arg.setAlwaysEncoded(false);
+            assertEquals("arg.getEncodedName() should be equal to name.?", "name.?", arg.getEncodedName());
+            assertEquals("arg.getValue() should be equal to value_here", "value_ here", arg.getEncodedValue());
+    
+            // When the name and value parameter does not have valid encoded value (Encoded
+            // and Decoded values are same)
+            // In real encoded value for "name.?" is "name.%3F" and the encoded value for
+            // "value_here" is "value_+here"
+            arg = new HTTPArgument("name.?", "value_ here", true, encoding);
+    
+            assertEquals("arg.getName() should be equal to name.?", "name.?", arg.getName());
+            assertEquals("arg.getValue() should be equal to value_here", "value_ here", arg.getValue());
+            assertEquals("arg.getEncodedName() should be equal to name.%3F", "name.%3F", arg.getEncodedName());
+            assertEquals("arg.getValue() should be equal to value_+here", "value_+here", arg.getEncodedValue());
+            arg.setAlwaysEncoded(false); // by default, name/value are encoded on fetch
+            assertEquals("arg.getEncodedName() should be equal to name.?", "name.?", arg.getEncodedName());
+            assertEquals("arg.getValue should() be equal to value_here", "value_ here", arg.getEncodedValue());
+    
+            // Try a real encoded argument
+            arg = new HTTPArgument("name.%3F", "value_+here", true, encoding);
+            assertEquals("arg.getName() should be equal to name.?", "name.?", arg.getName());
+            assertEquals("arg.getValue() should be equal to value_here", "value_ here", arg.getValue());
+            assertEquals("arg.getEncodedName() should be equal to name.%3F", "name.%3F", arg.getEncodedName());
+            assertEquals("arg.getValue() should be equal to value_+here", "value_+here", arg.getEncodedValue());
+            // Show that can bypass encoding:
+            arg.setAlwaysEncoded(false);
+            assertEquals("arg.getEncodedName() should be equal to name.?", "name.?", arg.getEncodedName());
```

Comment:
Again. The message does not clarify the intention of the test.
`should be equal to name.?` should be replaced with something like `setAlwaysEncoded=false => name.? should not be decoded` (I've no idea what setAlwaysEncoded does, and it is really hard to tell what the test is verifying)

Reply:
setAlwaysEncoded(param) method changes the property AlwaysEncoded of the HTTPArgument. It is then checked by the getEncodedName() and getEncodedValue() methods and if isAlwaysEncoded() returns true then Name/Value is encoded else it isn't encoded. Attaching the implementation of getEncodedName()

`    public String getEncodedName() {
        if (isAlwaysEncoded()) {
            return cache.getEncoded(getName());
        } else {
            return getName();
        }`

hence updated the message `setAlwaysEncoded=false => arg.getEncodedName() should not be encoded`